File: r/tests/testthat/test_competition_canvas_save_load.R
# test_competition_canvas_save_load.R — Round-trip tests for save/load_competition_canvas.

# ---------------------------------------------------------------------------
# Fixtures (reuse 1D and 2D trace helpers from sibling test files)
# ---------------------------------------------------------------------------

.voters_sl_1d <- c(-0.5, 0.1, 1.8)

.trace_sl_1d <- function() {
  competition_run(
    competitor_positions = c(0.0, 2.0),
    strategy_kinds       = c("sticker", "sticker"),
    voter_ideals         = .voters_sl_1d,
    dist_config          = make_dist_config(n_dims = 1L),
    engine_config        = make_competition_engine_config(
      seat_count  = 1L,
      seat_rule   = "plurality_top_k",
      max_rounds  = 2L,
      step_config = make_competition_step_config(kind = "fixed", fixed_step_size = 0.5)
    )
  )
}

.voters_sl_2d <- c(-0.5, 0.0,  0.0, 0.2,  1.8, 0.1)  # 3 voters, 2D flat

.trace_sl_2d <- function() {
  competition_run(
    competitor_positions = c(0.0, 0.0,  2.0, 0.0),
    strategy_kinds       = c("sticker", "sticker"),
    voter_ideals         = .voters_sl_2d,
    dist_config          = make_dist_config(n_dims = 2L),
    engine_config        = make_competition_engine_config(
      seat_count  = 1L,
      seat_rule   = "plurality_top_k",
      max_rounds  = 2L,
      step_config = make_competition_step_config(kind = "fixed", fixed_step_size = 0.5)
    )
  )
}

# ---------------------------------------------------------------------------
# 1D round-trip
# ---------------------------------------------------------------------------

test_that("save_competition_canvas writes a valid .scscanvas file (1D)", {
  skip_without_lib()
  trace <- .trace_sl_1d()
  w <- animate_competition_canvas(trace, voters = .voters_sl_1d, width = 900L, height = 400L)
  tmp <- tempfile(fileext = ".scscanvas")
  on.exit(unlink(tmp))

  result <- save_competition_canvas(w, tmp)

  expect_equal(result, tmp)
  expect_true(file.exists(tmp))
  raw <- jsonlite::read_json(tmp, simplifyVector = FALSE)
  expect_equal(raw$format, "scscanvas")
  expect_equal(raw$version, "1")
  expect_true(!is.null(raw$created))
  expect_true(startsWith(raw$generator, "socialchoicelab/r/"))
  expect_equal(raw$width,  900L)
  expect_equal(raw$height, 400L)
  expect_true("payload" %in% names(raw))
})

test_that("load_competition_canvas returns a widget with matching payload (1D)", {
  skip_without_lib()
  trace <- .trace_sl_1d()
  w_orig <- animate_competition_canvas(trace, voters = .voters_sl_1d, width = 900L, height = 400L)
  tmp <- tempfile(fileext = ".scscanvas")
  on.exit(unlink(tmp))
  save_competition_canvas(w_orig, tmp)

  w_loaded <- load_competition_canvas(tmp)

  expect_s3_class(w_loaded, "htmlwidget")
  expect_equal(w_loaded$x$n_dims,       w_orig$x$n_dims)
  expect_equal(w_loaded$x$voters_x,     w_orig$x$voters_x)
  expect_equal(w_loaded$x$rounds,       w_orig$x$rounds)
  expect_equal(w_loaded$x$positions,    w_orig$x$positions)
  expect_equal(w_loaded$x$vote_shares,  w_orig$x$vote_shares)
})

test_that("load_competition_canvas width/height override works (1D)", {
  skip_without_lib()
  trace <- .trace_sl_1d()
  w_orig <- animate_competition_canvas(trace, voters = .voters_sl_1d, width = 900L, height = 400L)
  tmp <- tempfile(fileext = ".scscanvas")
  on.exit(unlink(tmp))
  save_competition_canvas(w_orig, tmp)

  w_loaded <- load_competition_canvas(tmp, width = 1200L, height = 500L)

  expect_equal(w_loaded$width,  1200L)
  expect_equal(w_loaded$height, 500L)
})

# ---------------------------------------------------------------------------
# 2D round-trip
# ---------------------------------------------------------------------------

test_that("save_competition_canvas writes a valid .scscanvas file (2D)", {
  skip_without_lib()
  trace <- .trace_sl_2d()
  w <- animate_competition_canvas(trace, voters = .voters_sl_2d, width = 900L, height = 800L)
  tmp <- tempfile(fileext = ".scscanvas")
  on.exit(unlink(tmp))

  save_competition_canvas(w, tmp)

  raw <- jsonlite::read_json(tmp, simplifyVector = FALSE)
  expect_equal(raw$format, "scscanvas")
  expect_equal(raw$width,  900L)
  expect_equal(raw$height, 800L)
  expect_true("voters_x" %in% names(raw$payload))
  expect_true("ylim"     %in% names(raw$payload))
})

test_that("load_competition_canvas returns a widget with matching payload (2D)", {
  skip_without_lib()
  trace <- .trace_sl_2d()
  w_orig <- animate_competition_canvas(trace, voters = .voters_sl_2d, width = 900L, height = 800L)
  tmp <- tempfile(fileext = ".scscanvas")
  on.exit(unlink(tmp))
  save_competition_canvas(w_orig, tmp)

  w_loaded <- load_competition_canvas(tmp)

  expect_s3_class(w_loaded, "htmlwidget")
  expect_equal(w_loaded$x$voters_x,     w_orig$x$voters_x)
  expect_equal(w_loaded$x$ylim,         w_orig$x$ylim)
  expect_equal(w_loaded$x$positions,    w_orig$x$positions)
  expect_equal(w_loaded$x$vote_shares,  w_orig$x$vote_shares)
})

# ---------------------------------------------------------------------------
# Error cases
# ---------------------------------------------------------------------------

test_that("save_competition_canvas errors on non-widget input", {
  expect_error(
    save_competition_canvas(list(x = 1), tempfile()),
    "must be an htmlwidget"
  )
})

test_that("load_competition_canvas errors on missing file", {
  expect_error(
    load_competition_canvas(tempfile(fileext = ".scscanvas")),
    "file not found"
  )
})

test_that("load_competition_canvas errors on wrong format", {
  skip_without_lib()
  tmp <- tempfile(fileext = ".scscanvas")
  on.exit(unlink(tmp))
  jsonlite::write_json(list(format = "other", payload = list()), tmp, auto_unbox = TRUE)
  expect_error(
    load_competition_canvas(tmp),
    "unexpected format"
  )
})

File: python/tests/test_competition_canvas_save_load.py
# test_competition_canvas_save_load.py — Round-trip tests for save/load_competition_canvas.

import json
import tempfile
from pathlib import Path

import numpy as np

import socialchoicelab as scl
import socialchoicelab.plots as sclp

# ---------------------------------------------------------------------------
# Fixtures
# ---------------------------------------------------------------------------

_VOTERS_1D = np.array([[-0.5], [0.1], [1.8]])
_COMPS_1D  = np.array([[0.0], [2.0]])
_ENGINE_1D = scl.CompetitionEngineConfig(
    seat_count=1,
    seat_rule="plurality_top_k",
    max_rounds=2,
    step_config=scl.CompetitionStepConfig(kind="fixed", fixed_step_size=0.5),
)
_DIST_1D = scl.DistanceConfig(salience_weights=[1.0])

_VOTERS_2D = np.array([[-0.5, 0.0], [0.0, 0.2], [1.8, 0.1]])
_COMPS_2D  = np.array([[0.0, 0.0], [2.0, 0.0]])
_DIST_2D   = scl.DistanceConfig(salience_weights=[1.0, 1.0])
_ENGINE_2D = scl.CompetitionEngineConfig(
    seat_count=1,
    seat_rule="plurality_top_k",
    max_rounds=2,
    step_config=scl.CompetitionStepConfig(kind="fixed", fixed_step_size=0.5),
)


def _trace_1d():
    return scl.competition_run(
        _COMPS_1D, ["sticker", "sticker"], _VOTERS_1D,
        dist_config=_DIST_1D, engine_config=_ENGINE_1D,
    )


def _trace_2d():
    return scl.competition_run(
        _COMPS_2D, ["sticker", "sticker"], _VOTERS_2D,
        dist_config=_DIST_2D, engine_config=_ENGINE_2D,
    )


# ---------------------------------------------------------------------------
# 1D round-trip via payload_path
# ---------------------------------------------------------------------------

def test_payload_path_writes_scscanvas_file_1d():
    trace = _trace_1d()
    with tempfile.NamedTemporaryFile(suffix=".scscanvas", delete=False) as f:
        tmp = f.name
    try:
        sclp.animate_competition_canvas(
            trace, voters=_VOTERS_1D.ravel(), width=900, height=400,
            payload_path=tmp,
        )
        assert Path(tmp).exists()
        envelope = json.loads(Path(tmp).read_text())
        assert envelope["format"] == "scscanvas"
        assert envelope["version"] == "1"
        assert "created" in envelope
        assert envelope["generator"].startswith("socialchoicelab/python/")
        assert envelope["width"] == 900
        assert envelope["height"] == 400
        assert "payload" in envelope
    finally:
        Path(tmp).unlink(missing_ok=True)


def test_load_competition_canvas_returns_html_1d():
    trace = _trace_1d()
    with tempfile.NamedTemporaryFile(suffix=".scscanvas", delete=False) as f:
        tmp = f.name
    try:
        html_orig = sclp.animate_competition_canvas(
            trace, voters=_VOTERS_1D.ravel(), width=900, height=400,
            payload_path=tmp,
        )
        html_loaded = sclp.load_competition_canvas(tmp)
        assert "<!DOCTYPE html>" in html_loaded
        # Payload content is identical so the serialised JSON inside should match.
        envelope = json.loads(Path(tmp).read_text())
        payload = envelope["payload"]
        assert payload["voters_x"] == json.loads(
            html_orig[html_orig.index("HTMLWidgets._init(") :]
            .split("HTMLWidgets._init(")[0]  # not used; just verify voters_x via envelope
            or "{}"
        ).get("voters_x", payload["voters_x"])
        # Key structure check: loaded HTML contains the same voter count.
        assert str(len(payload["voters_x"])) in html_loaded
    finally:
        Path(tmp).unlink(missing_ok=True)


def test_load_competition_canvas_payload_matches_original_1d():
    trace = _trace_1d()
    with tempfile.NamedTemporaryFile(suffix=".scscanvas", delete=False) as f:
        tmp = f.name
    try:
        sclp.animate_competition_canvas(
            trace, voters=_VOTERS_1D.ravel(), width=900, height=400,
            payload_path=tmp,
        )
        envelope = json.loads(Path(tmp).read_text())
        payload = envelope["payload"]
        assert payload["n_dims"] == 1
        assert len(payload["voters_x"]) == 3
        assert len(payload["rounds"]) == 3   # Round 1, Round 2, Final
        assert len(payload["positions"]) == 3
    finally:
        Path(tmp).unlink(missing_ok=True)


# ---------------------------------------------------------------------------
# 2D round-trip via payload_path
# ---------------------------------------------------------------------------

def test_payload_path_writes_scscanvas_file_2d():
    trace = _trace_2d()
    with tempfile.NamedTemporaryFile(suffix=".scscanvas", delete=False) as f:
        tmp = f.name
    try:
        sclp.animate_competition_canvas(
            trace, voters=_VOTERS_2D, width=900, height=800,
            payload_path=tmp,
        )
        envelope = json.loads(Path(tmp).read_text())
        assert envelope["format"] == "scscanvas"
        assert envelope["width"] == 900
        assert envelope["height"] == 800
        payload = envelope["payload"]
        assert "voters_x" in payload
        assert "voters_y" in payload
        assert len(payload["voters_x"]) == 3
    finally:
        Path(tmp).unlink(missing_ok=True)


def test_load_competition_canvas_returns_html_2d():
    trace = _trace_2d()
    with tempfile.NamedTemporaryFile(suffix=".scscanvas", delete=False) as f:
        tmp = f.name
    try:
        sclp.animate_competition_canvas(
            trace, voters=_VOTERS_2D, width=900, height=800,
            payload_path=tmp,
        )
        html_loaded = sclp.load_competition_canvas(tmp)
        assert "<!DOCTYPE html>" in html_loaded
        assert "competition-canvas" in html_loaded
    finally:
        Path(tmp).unlink(missing_ok=True)


def test_load_competition_canvas_width_height_override():
    trace = _trace_2d()
    with tempfile.NamedTemporaryFile(suffix=".scscanvas", delete=False) as f:
        tmp = f.name
    try:
        sclp.animate_competition_canvas(
            trace, voters=_VOTERS_2D, width=900, height=800,
            payload_path=tmp,
        )
        html = sclp.load_competition_canvas(tmp, width=1200, height=600)
        assert "1200px" in html
        assert "600px" in html
    finally:
        Path(tmp).unlink(missing_ok=True)


# ---------------------------------------------------------------------------
# top-level import
# ---------------------------------------------------------------------------

def test_load_competition_canvas_importable_from_top_level():
    from socialchoicelab import load_competition_canvas  # noqa: F401
    assert callable(load_competition_canvas)


# ---------------------------------------------------------------------------
# Error cases
# ---------------------------------------------------------------------------

def test_load_competition_canvas_missing_file():
    import pytest
    with pytest.raises(FileNotFoundError, match="file not found"):
        sclp.load_competition_canvas("/tmp/does_not_exist_xyz.scscanvas")


def test_load_competition_canvas_wrong_format():
    import pytest
    with tempfile.NamedTemporaryFile(
        suffix=".scscanvas", mode="w", delete=False, encoding="utf-8"
    ) as f:
        json.dump({"format": "other", "payload": {}}, f)
        tmp = f.name
    try:
        with pytest.raises(ValueError, match="unexpected format"):
            sclp.load_competition_canvas(tmp)
    finally:
        Path(tmp).unlink(missing_ok=True)
